compiler-lib/src/sem.rs

```diff
@@ -1,6 +1,163 @@
+    //  - type checking of expressions
+    SecondPass::visit(ast, context, &classes_and_members)
+}
+
+struct ClassesAndMembersVisitor;
+
+impl ClassesAndMembersVisitor {
+    pub fn visit<'a, 'f>(
+        ast: &'a ast::AST<'f>,
+        context: &context::Context<'_>,
+    ) -> Result<ClassesAndMembers<'a, 'f>, Error> {
+        let mut res = HashMap::new();
+        Self::do_visit(&NodeKind::from(ast), context, &mut res)?;
+        Ok(res)
+    }
+
+    fn do_visit<'a, 'f>(
+        n: &NodeKind<'a, 'f>,
+        _context: &context::Context<'_>,
+        res: &mut ClassesAndMembers<'a, 'f>,
+    ) -> Result<(), Error> {
+        use self::{ast, NodeKind::*};
+        match n {
+            AST(ast::AST::Program(p)) => {
+                for class in &p.data.classes {
+                    for member in class.members.iter() {
+                        let discr = ast::ClassMemberKindDiscriminants::from(&member.kind);
+                        res.insert(
+                            (Rc::clone(&class.name), Rc::clone(&member.name), discr),
+                            &member,
+                        );
+                    }
+                }
+                Ok(())
+            }
+            x => {
+                return Err(format_err!(
+                    "unexpected ast node passed to first pass: {:?}",
+                    x
+                ))
+            }
+        }
+    }
+}
+
+type ScopeDefs<T> = HashMap<Symbol, T>;
+
+struct Scope<T> {
+    parent: Option<Rc<Scope<T>>>,
+    definitions: ScopeDefs<T>,
+}
+
+impl<T> Scope<T> {
+    pub fn root() -> Rc<Self> {
+        Rc::new(Scope {
+            parent: None,
+            definitions: ScopeDefs::new(),
+        })
+    }
+    pub fn enter_scope(parent: Rc<Self>) -> Self {
+        Scope {
+            parent: Some(parent),
+            definitions: ScopeDefs::new(),
+        }
+    }
+    pub fn leave(self) -> Option<Rc<Self>> {
+        self.parent
+    }
+    pub fn define(&mut self, sym: &Symbol, val: T) -> Result<(), ()> {
+        use std::collections::hash_map::Entry;
+        let key = Rc::clone(sym);
+        match self.definitions.entry(key) {
+            Entry::Occupied(o) => return Err(()),
+            Entry::Vacant(e) => e.insert(val),
+        };
+        Ok(())
+    }
+}
+
+struct SecondPass;
+
+// TODO proper enum, but Type is definitely necessary
+type VarDef<'a, 'f> = &'a Spanned<'f, ast::Type>;
+struct TyDef; // TODO
+struct MethDef; // TODO
+
+impl SecondPass {
+    pub fn visit<'a, 'f>(
+        ast: &'a ast::AST<'f>,
+        context: &context::Context<'_>,
+        _classes_and_members: &ClassesAndMembers<'a, 'f>,
+    ) -> Result<(), Error> {
+        // TODO java.lang.System.out scope
+        // TODO ontop of that, scope of this package
+        let mut vardefs: Rc<Scope<VarDef<'a, 'f>>> = Scope::root();
+        let mut tydefs: Rc<Scope<TyDef>> = Scope::root();
+        let mut methdefs: Rc<Scope<MethDef>> = Scope::root();
```

Comment:
Put all this data structures in the `SecondPass` struct, make the `visit` method a `new` function.

Then call 
```rust
let visitor = SecondPass::new(cx)
visitor.do_visit(ast)
```
from the check function.

Same goes for FirstPass and should be our standard approach on implementing visitors IMO. That way a visitor doesn't need as many parameters as data it has to hold.